atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/KvTableMappingService.java

```diff
@@ -110,35 +130,136 @@ private TableReference getAlreadyExistingMappedTableName(TableReference tableRef
+
+        // Need to invalidate the table refs in case we end up re-creating the same table again.
+        updateTableMapWith(() -> tableMapWithInvalidatedTableRefs(tableRefs));
+    }
+
+    private BiMap<TableReference, TableReference> tableMapWithInvalidatedTableRefs(Set<TableReference> tableRefs) {
+        BiMap<TableReference, TableReference> newMap = HashBiMap.create(tableMap.get());
+        tableRefs.forEach(newMap::remove);
+        return newMap;
+    }
+
+    @Override
+    public TableReference getMappedTableName(TableReference tableRef) throws TableMappingNotFoundException {
+        if (tableRef.getNamespace().isEmptyNamespace()) {
+            return tableRef;
+        }
+
+        TableReference shortTableName = getMappedTableRef(tableRef);
+        validateShortName(tableRef, shortTableName);
+        return shortTableName;
+    }
+
+    private TableReference getMappedTableRef(TableReference tableRef)
+            throws TableMappingNotFoundException {
+        TableReference candidate = tableMap.get().get(tableRef);
+        if (candidate == null) {
+            updateTableMap();
+            candidate = tableMap.get().get(tableRef);
+            if (candidate == null) {
+                throw new TableMappingNotFoundException("Unable to resolve mapping for table reference " + tableRef);
+            }
+        }
+        return candidate;
+    }
+
+    protected void validateShortName(TableReference tableRef, TableReference shortName) {
+        Validate.isTrue(Schemas.isTableNameValid(shortName.getQualifiedName()),
+                "Table mapper has an invalid table name for table reference %s: %s", tableRef, shortName);
+    }
+
+
+    @Override
+    public <T> Map<TableReference, T> mapToShortTableNames(Map<TableReference, T> toMap)
+            throws TableMappingNotFoundException {
+        Map<TableReference, T> newMap = Maps.newHashMap();
+        for (Map.Entry<TableReference, T> e : toMap.entrySet()) {
+            newMap.put(getMappedTableName(e.getKey()), e.getValue());
         }
-        // Need to invalidate the table ref in case we end up re-creating the same table
-        // again. Frequently when we drop one table we end up dropping a bunch of tables,
-        // so just invalidate everything.
-        tableMap.set(HashBiMap.<TableReference, TableReference>create());
+        return newMap;
     }
 
     @Override
+    public Map<TableReference, TableReference> generateMapToFullTableNames(Set<TableReference> tableRefs) {
+        Map<TableReference, TableReference> shortNameToFullTableName = Maps.newHashMapWithExpectedSize(
+                tableRefs.size());
+        Set<TableReference> tablesToReload = Sets.newHashSet();
+        for (TableReference inputName : tableRefs) {
+            if (inputName.isFullyQualifiedName()) {
+                shortNameToFullTableName.put(inputName, inputName);
+            } else {
+                TableReference candidate = tableMap.get().inverse().get(inputName);
```

Comment:
similarly here

Reply:
The fix here is simple, so although it adds some overhead, makes sense